ucp/_libs/core.pyx

```diff
@@ -175,13 +175,26 @@ async def listener_handler(endpoint, ctx, func, guarantee_msg_order):
     del ep
     del ctx
 
-    # Finally, we call `func` asynchronously (even if it isn't coroutine)
+    # Finally, we call `func`
     if asyncio.iscoroutinefunction(func):
         await func(pub_ep)
     else:
-        async def _func(ep):  # coroutine wrapper
-            func(ep)
-        await _func(pub_ep)
+        func(pub_ep)
+
+
+def application_context_finalizer(children, worker, context, epoll_fd):
+    """
+    Finalizer function for `ApplicationContext` object, which is
+    more reliable than __dealloc__.
+    """
+    for child in children:
+        child = child()
```

Comment:
Should we disambiguate the name here somehow? It's not totally clear (at least not to me) what's happening

Reply:
Good point, will fix it :)

Reply:
Furthermore as part of the implementation of https://github.com/rapidsai/ucx-py/pull/418, I will streamline the cleanup processes.